diffrax/_solver/srk.py

```diff
@@ -0,0 +1,624 @@
+                    add_levy_to_w(a_w_j, *a_levy_list_j),  # type: ignore
+                    w_g,
+                    *levy_gs_list,
+                )
+            else:
+                # carry = (j, _tfs, _wgs, _levy_gs_list) where
+                # _tfs = Array[tf1, tf2, ..., tf{j-1}, 0, 0, ..., 0]
+                # _wgs = Array[wg1, wg2, ..., wg{j-1}, 0, 0, ..., 0]
+                # _levy_gs_list = [H_gs, K_gs] (if those entries exist) where
+                # H_gs = Array[Hg1, Hg2, ..., Hg{j-1}, 0, 0, ..., 0]
+                # K_gs = Array[Kg1, Kg2, ..., Kg{j-1}, 0, 0, ..., 0]
+                # tfi = drift.vf_prod(t0 + c_i*h, y_i, args, h) (i.e. hki = h * k_i)
+                # wgi = diffusion.vf_prod(t0 + c_i*h, y_i, args, w) (i.e. wgi = w * g_i)
+
+                wg_sum = sum_prev_stages(_wgs, a_w_j)
+                levy_sum_list = [
+                    sum_prev_stages(levy_gs, a_levy_j)
+                    for a_levy_j, levy_gs in zip(a_levy_list_j, _levy_gs_list)
+                ]
+
+                _diffusion_result = jtu.tree_map(
+                    lambda _wg, *_levy_g: sum(_levy_g, _wg), wg_sum, *levy_sum_list
+                )
+
+            # compute Σ_{i=1}^{j-1} a_j_i tf_i
+            _drift_result = sum_prev_stages(_tfs, a_j)
+
+            # z_j = y_0 + h (Σ_{i=1}^{j-1} a_j_i k_i) + g * (a_w_j * ΔW + cH_j * ΔH)
+            z_j = (y0**ω + _drift_result**ω + _diffusion_result**ω).ω
+
+            def insert_jth_stage(results, res_j):
+                return jtu.tree_map(
+                    lambda res_leaf, j_leaf: res_leaf.at[j].set(j_leaf), results, res_j
+                )
+
+            def compute_and_insert_tfj(_tfs_in):
+                tf_j = drift.vf_prod(t0 + c_j * h, z_j, args, h)
+                return insert_jth_stage(_tfs_in, tf_j)
+
+            if self.tableau.ignore_stage_f is None:
+                _tfs = compute_and_insert_tfj(_tfs)
+            else:
+                drift_pred = jnp.logical_not(ignore_stage_f[j])
+                _tfs = lax.cond(drift_pred, compute_and_insert_tfj, lambda x: x, _tfs)
+
+            if additive_noise:
+                return (j + 1, _tfs, None, None), None
+
+            def compute_and_insert_gj(_wgs_in, _levy_gs_list_in):
+                wg_j = diffusion.vf_prod(t0 + c_j * h, z_j, args, w)
+                new_wgs = insert_jth_stage(_wgs_in, wg_j)
+
+                levy_g_list_j = [
+                    diffusion.vf_prod(t0 + c_j * h, z_j, args, levy)
+                    for levy in levy_areas
+                ]
+                new_levy_gs_list = insert_jth_stage(_levy_gs_list_in, levy_g_list_j)
+                return new_wgs, new_levy_gs_list
+
+            if self.tableau.ignore_stage_g is None:
+                _wgs, _levy_gs_list = compute_and_insert_gj(_wgs, _levy_gs_list)
+            else:
+                diffusion_pred = jnp.logical_not(ignore_stage_g[j])
+                _wgs, _levy_gs_list = lax.cond(
+                    diffusion_pred,
```

Comment:
Recall that if the condition of a `lax.cond` is batched, then the `cond` will turn into a `jnp.where` and both branches will be evaluated. Now of course in practice, this should never happen here -- the condition should depend only on the counter of the scan, which shouldn't depend on any batched inputs and so shouldn't pick up any batch tracers.

To assert that this is defintiely the case, I've got a secret weapon: by writing `lax.cond(eqxi.nonbatchable(diffusion_pred), ...)`, then we will explicitly raise an error (rather than silently converting to a `jnp.where`) if the predicate is in fact batched.

(This is one of the many hidden tools that I've got squirreled away in the `eqxi` namespace... !)

Reply:
That's great! So my thinking was, sometimes if it was batched we cannot avoid doing the computation anyway, but at least in non-batched cases we could speed it up (and I suspected there shouldn't be any batching over the tableau anyway). But yes, I agree it is better to make this explicit.